### DISCUSSION TITLE
How / where do I start a new module?

### DISCUSSION THREAD

**Discussion by ndemarco** [Tecnativa/doodba#639] [Doodba - The template] [answered]
In [Daily Usage ](https://github.com/Tecnativa/doodba-copier-template/blob/main/docs/daily-usage.md), my first use after starting the dev environment and getting comfy in the VSCode workspace is to start developing a new module. Doodba is self-described as opinionated 😜. As a new Odoo developer, how and where do I start writing my first new custom addon module?

It seems like addons should go in odoo/custom/src/. Confirm I can a) create an addon module directly in odoo/custom/src/ or b) I can create a git-managed addon and reference it for `git-aggregator`. How do I do each of these?

Might sound basic, but it's the first thing I think people will want to do when spinning a new Odoo instance. Or maybe new Odoo-ers should not start with Doodba and just spin up a local Odoo docker container and start there?

**Comment by pedrobaeza** [comment]
The `private` folder is there for hosting your custom modules. Place them there.

**Comment by ndemarco** [comment]
So put non-git-aggregator-controlled modules in `odoo/custom/src/private` folder?

I successfully added OCA modules by including them in `odoo/custom/src/addons.yaml`, but that is magic from the [Doodba wizard](https://github.com/Tecnativa/doodba?tab=readme-ov-file#optodoocustomsrcreposyaml) that is not currently clear.

My question is: how do I add git controlled modules that are not part of OCA, or even in GitHub - like from local git-managed directories.

**Reply by pedrobaeza** [reply]
That "magic" is explained here AFAIK: https://github.com/Tecnativa/doodba?tab=readme-ov-file#automatic-download-of-repos

For adding other repositories that are not OCA ones, you have to add the corresponding entries in repos.yaml, with the remotes. This is not something of Doodba itself, but https://github.com/acsone/git-aggregator. Check there for some more info. You can even add other GitHub/Gitlab private repositories if you add deploy tokens in the remote URL, or you add SSH keys properly. Check other questions where this has been already asked.

**Reply by ap-wtioit** [reply, marked as answer]
If you can add it with `git remote add name url` you can add it in the repos.yaml:
```yaml
myrepo:
  defaults:
    # This will use git shallow clones.
    # $DEPTH_DEFAULT is 1 in test and prod, but 100 in devel.
    # $DEPTH_MERGE is always 100.
    # You can use any integer value, OTOH.
    depth: $DEPTH_MERGE
  remotes:
    name: url
  target: name $ODOO_VERSION
  merges:
    - name $ODOO_VERSION
```

What do you mean with local git-managed? A local gitlab instance of a directory?

And it is highly recommend to use a branch named $ODOO_VERSION (e.g. 18.0) but you can also use any custom ref name in target. (But be careful, git-aggregate performs `git checkout -B` with this name, so best do not use anything like name/somebranch or refs/...)

**Reply by ndemarco** [reply]
@pedrobaeza thanks for explaining. It's starting to make sense.

> Check other questions where this has been already asked.

I've been searching, but clearly have not been using the right search terms.

**Reply by ndemarco** [reply]
> What do you mean with local git-managed? A local gitlab instance of a directory?

Not really. I mean a local git-managed folder (~/my_throwaway_module/.git).

> And it is highly recommend to use a branch named $ODOO_VERSION (e.g. 18.0) but you can also use any custom ref name in target. (But be careful, git-aggregate performs `git checkout -B` with this name, so best do not use anything like name/somebranch or refs/...)

Good point - makes sense.

**Comment by ndemarco** [comment]
I appended to the default `repos.yaml` like this:

```
# See https://github.com/Tecnativa/doodba#optodoocustomsrcreposyaml
./odoo:
  defaults:
    # Shallow repositories ($DEPTH_DEFAULT=1) are faster & thinner
    # You may need a bigger depth when merging PRs (use $DEPTH_MERGE
    # for a sane value of 100 commits)
    depth: $DEPTH_DEFAULT
  remotes:
    ocb: https://github.com/OCA/OCB.git
    odoo: https://github.com/odoo/odoo.git
    openupgrade: https://github.com/OCA/OpenUpgrade.git
  target: ocb $ODOO_VERSION
  merges:
    - ocb $ODOO_VERSION
    # Example of a merge of the PR with the number <PR>
    # - oca refs/pull/<PR>/head
./my_library:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    nickydoes: https://github.com/ndemarco/warp-my_library.git
  target: nickydoes $ODOO_VERSION
  merges:
    - nickydoes $ODOO_VERSION
 ```

and ran `invoke git-aggregate`

The repo was added to `opt/odoo/custom/src`, so I had `opt/odoo/custom/src/my_library` which is not on the addons path. I `mv`'d it to `/opt/odoo/custom/src/private`, Updated the Apps List, and the module was listed (as expected).

Why did my custom module get aggregated into the `src` dir?

Aggregate logs:
```
nick@Nix-XPS-XIII:~/projects/warp/odoo/custom/src/private$ invoke git-aggregate
Reinitialized existing Git repository in /home/nick/projects/warp/.git/
pre-commit installed at .git/hooks/pre-commit
time="2024-12-19T13:30:12-05:00" level=warning msg="/home/nick/projects/warp/setup-devel.yaml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion"
 Network warp_public  Creating
 Network warp_public  Created
doodba INFO: Running gitaggregate with /opt/odoo/custom/src/repos.yaml
(INFO) [18:30:13] git_aggregator.repo  odoo   Start aggregation of /opt/odoo/custom/src/odoo
(INFO) [18:30:13] git_aggregator.repo  odoo   Switch to branch 17.0
(INFO) [18:30:13] git_aggregator.repo  my_library   Start aggregation of /opt/odoo/custom/src/my_library
(INFO) [18:30:13] git_aggregator.repo  my_library   Cloning git repository https://github.com/ndemarco/warp-my_library.git in /opt/odoo/custom/src/my_library
Reset branch '17.0'
Your branch is up to date with 'origin/17.0'.
(INFO) [18:30:13] git_aggregator.repo  odoo   Remote already exists ocb <https://github.com/OCA/OCB.git>
(INFO) [18:30:13] git_aggregator.repo  odoo   Remote already exists odoo <https://github.com/odoo/odoo.git>
(INFO) [18:30:13] git_aggregator.repo  odoo   Remote already exists openupgrade <https://github.com/OCA/OpenUpgrade.git>
(INFO) [18:30:13] git_aggregator.repo  odoo   Fetching required remotes
Cloning into '/opt/odoo/custom/src/my_library'...
From https://github.com/OCA/OCB
 * branch              17.0       -> FETCH_HEAD
(INFO) [18:30:14] git_aggregator.repo  odoo   Checking repo status
(INFO) [18:30:14] git_aggregator.repo  odoo   Reset branch to ocb 17.0
(INFO) [18:30:14] git_aggregator.repo  my_library   Switch to branch 17.0
Reset branch '17.0'
Your branch is up to date with 'origin/17.0'.
(INFO) [18:30:14] git_aggregator.repo  my_library   Adding remote nickydoes <https://github.com/ndemarco/warp-my_library.git>
(INFO) [18:30:14] git_aggregator.repo  my_library   Fetching required remotes
From https://github.com/ndemarco/warp-my_library
 * branch            17.0       -> FETCH_HEAD
 * [new branch]      17.0       -> nickydoes/17.0
(INFO) [18:30:14] git_aggregator.repo  my_library   Checking repo status
(INFO) [18:30:14] git_aggregator.repo  my_library   Reset branch to nickydoes 17.0
(INFO) [18:30:14] git_aggregator.repo  odoo   Execute shell after commands
(INFO) [18:30:14] git_aggregator.repo  odoo   End aggregation of /opt/odoo/custom/src/odoo
(INFO) [18:30:14] git_aggregator.repo  my_library   Execute shell after commands
(INFO) [18:30:14] git_aggregator.repo  my_library   End aggregation of /opt/odoo/custom/src/my_library
```

**Reply by ap-wtioit** [reply]
my_library should have been added to the addons path / all modules in it (that are defined in addons.yaml) should be symlinked to /opt/odoo/auto/addons/{addon_name} upon odoo container start (unless your entrypoint(s) are not working, or you redefined the addons path)

**Reply by ndemarco** [reply]
Thanks for that. I'm still looking for why, but I recreated a new project, starting with `copier copy gh:Tecnativa/doodba-copier-template project_name`. I added the my public `my_library` repo to `repos.yaml`, then `invoke develop img-build git-aggregate`.

my_library was again added to `opt\odoo\custom\src`.

edit: I tried once more after a full clearing out `docker system prune -a --volumes`. Same result:
- `my_library` is in `src` but not symlinked in `addons`. Weird.


Refresher:The my_library addition to `repos.yaml`:
```
./my_library:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    nickydoes: https://github.com/ndemarco/warp-my_library.git
  target: nickydoes $ODOO_VERSION
  merges:
    - nickydoes $ODOO_VERSION
 ```

**Reply by ap-wtioit** [reply]
from your github repo (branch 18.0) it looks like you have the addon directly in the repo (`__manifest__.py`) in main directory. Usually we work with repos that have addons inside them and are not addons by themselves: https://github.com/OCA/web/tree/18.0

you can try if git-aggregator/doodba supports the repo being an addon with:
repos.yaml (for git-aggregator)
```yaml
./ndemarco/my_library:
  defaults:
    depth: $DEPTH_DEFAULT
  remotes:
    nickydoes: https://github.com/ndemarco/warp-my_library.git
  target: nickydoes $ODOO_VERSION
  merges:
    - nickydoes $ODOO_VERSION
```
addons.yaml (for doodba entrypoint)
```
./ndemarco:
    - "my_library"
```

**But** i would advice you (if you have the option) to organize your repos and plugins the same way the OCA does. (And odoo enterprise also works this way)
